les/fetcher.go

```diff
@@ -66,6 +75,7 @@ type fetcherPeerInfo struct {
 	bestConfirmed       *fetcherTreeNode
 	nodeByHash          map[common.Hash]*fetcherTreeNode
 	firstUpdateStats    *updateStatsEntry
+	lock                sync.RWMutex
```

Comment:
Why do you need all these locks? Also, are you sure this is safe? The light fetcher is not a performance critical part of the code, lightFetcher.lock did protect everything, it was the safest and simplest locking scheme.

Reply:
Yes, it fixes datarace  at checkUpdateStats.

Reply:
Which was introduced in unregisterPeer, see below. This lock is not needed.

Reply:
```
WARNING: DATA RACE
Read at 0x00c4233297d8 by goroutine 172:
  github.com/ethereum/go-ethereum/les.(*lightFetcher).checkUpdateStats()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/fetcher.go:948 +0x277
  github.com/ethereum/go-ethereum/les.(*lightFetcher).unregisterPeer()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/fetcher.go:248 +0xae
  github.com/ethereum/go-ethereum/les.(*peerSet).Unregister()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/peer.go:601 +0x246
  github.com/ethereum/go-ethereum/les.testOdr()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/odr_test.go:206 +0x8e2
  github.com/ethereum/go-ethereum/les.TestOdrGetBlockLes2()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/odr_test.go:44 +0x56
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Previous write at 0x00c4233297d8 by goroutine 223:
  github.com/ethereum/go-ethereum/les.(*lightFetcher).checkAnnouncedHeaders()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/fetcher.go:703 +0x48a
  github.com/ethereum/go-ethereum/les.(*lightFetcher).newHeaders()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/fetcher.go:627 +0x16c
  github.com/ethereum/go-ethereum/les.(*lightFetcher).syncLoop()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/fetcher.go:221 +0x4ff

```